Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp

```diff
@@ -0,0 +1,238 @@
+  m_interrupt = true;
+  // Wait until the function has completed. The background thread will exit
+  // automatically
+  while (m_capturing) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  };
+}
+
+/**
+ * Check if there is data available to extract
+ * @return True if data has been accumulated so that extractData()
+ * can be called, false otherwise
+ */
+bool KafkaHistoStreamDecoder::hasData() const {
+  std::lock_guard<std::mutex> lock(m_buffer_mutex);
+  return !m_buffer.empty();
+}
+
+/**
+ * Check for an exception thrown by the background thread and rethrow
+ * it if necessary.
+ * @return A pointer to the data collected since the last call to this
+ * method
+ */
+API::Workspace_sptr KafkaHistoStreamDecoder::extractData() {
+  if (m_exception) {
+    throw *m_exception;
+  }
+
+  auto workspace_ptr = extractDataImpl();
+  return workspace_ptr;
+}
+
+// -----------------------------------------------------------------------------
+// Private members
+// -----------------------------------------------------------------------------
+
+API::Workspace_sptr KafkaHistoStreamDecoder::extractDataImpl() {
+  std::lock_guard<std::mutex> lock(m_buffer_mutex);
+
+  if (!m_capturing) {
+    throw Exception::NotYet("Local buffers not initialized.");
+  }
+
+  if (m_buffer.empty()) {
+    throw Exception::NotYet("No message to process yet.");
+  }
+
+  auto histoMsg = GetEventHistogram(m_buffer.c_str());
+
+  auto shape = histoMsg->current_shape();
+  auto nbins = shape->Get(0) - 1;
+  auto nspectra = static_cast<size_t>(shape->Get(1));
+
+  auto metadata = histoMsg->dim_metadata();
+  auto metadimx = metadata->GetAs<DimensionMetaData>(0);
+  auto metadimy = metadata->GetAs<DimensionMetaData>(1);
+  auto xbins = metadimx->bin_boundaries_as_ArrayDouble()->value();
+
+  // Compiler warnings if one tries to use xbins->begin()/end()
+  auto *bindata = xbins->data();
+  HistogramData::BinEdges binedges(&bindata[0], &bindata[xbins->size()]);
+
+  API::MatrixWorkspace_sptr ws{DataObjects::create<DataObjects::Workspace2D>(
+      *m_workspace, nspectra, binedges)};
+
+  ws->setIndexInfo(m_workspace->indexInfo());
+  auto data = histoMsg->data_as_ArrayDouble()->value();
+
+  // Set the units
+  ws->getAxis(0)->setUnit(metadimx->unit()->c_str());
+  ws->setYUnit(metadimy->unit()->c_str());
+
+  std::vector<double> counts;
+  for (size_t i = 0; i < nspectra; ++i) {
+    const double *start = data->data() + (i * nbins);
+    counts.assign(start, start + nbins);
+    ws->setCounts(i, counts);
+  }
+
+  return ws;
+}
+
+/**
+ * Start decoding data from the streams into the internal buffers.
+ * Implementation designed to be entry point for new thread of execution.
+ * It catches all thrown exceptions.
+ */
+void KafkaHistoStreamDecoder::captureImpl() {
+  m_capturing = true;
+  try {
+    captureImplExcept();
+  } catch (std::exception &exc) {
+    m_exception = boost::make_shared<std::runtime_error>(exc.what());
```

Comment:
Can't this be a `std::unique_ptr` or a `boost::optional<std::runtime_error>` can't see what sharing has to do with this.